liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -0,0 +1,889 @@
+        
+        util.watchFiles(pom, outputDirectory, testOutputDirectory, executor, artifactPaths, configFile);
+    }
+
+    private void addArtifacts(org.eclipse.aether.graph.DependencyNode root, List<File> artifacts) {
+        if (root.getArtifact() != null) {
+            artifacts.add(root.getArtifact().getFile());
+        }
+
+        for (org.eclipse.aether.graph.DependencyNode node : root.getChildren()) {
+            addArtifacts(node, artifacts);
+        }
+    }
+
+    private List<Artifact> getNewDependencies(List<Dependency> dependencies, List<Dependency> existingDependencies)
+            throws MojoExecutionException {
+        List<Artifact> updatedArtifacts = new ArrayList<Artifact>();
+        for (Dependency dep : dependencies) {
+            boolean newDependency = true;
+            for (Dependency existingDep : existingDependencies) {
+                if (dep.getArtifactId().equals(existingDep.getArtifactId())) {
+                    newDependency = false;
+                }
+            }
+            if (newDependency) {
+                log.debug("New dependency found: " + dep.getArtifactId());
+                Artifact artifact = getArtifact(dep.getGroupId(), dep.getArtifactId(), dep.getType(), dep.getVersion());
+                updatedArtifacts.add(artifact);
+            }
+        }
+        return updatedArtifacts;
+    }
+    
+    private String readConfigurations(String groupId, String artifactId, String id, String attribute) {
+        Plugin plugin = project.getPlugin(groupId + ":" + artifactId);
+        if (plugin == null) {
+            plugin = plugin(groupId(groupId), artifactId(artifactId), version("RELEASE"));
+        }
+        Xpp3Dom config = null;
+
+        List<Plugin> buildPlugins = project.getBuildPlugins();
+        for (Plugin p : buildPlugins) {
+            if (p.equals(plugin)) {
+                config = (Xpp3Dom) p.getConfiguration();
+                
+                PluginExecution pe;
+                Map<String, PluginExecution> peMap = p.getExecutionsAsMap();
+                
+                if ((pe = peMap.get(id)) != null ){
+                    Xpp3Dom executionConfig = (Xpp3Dom) pe.getConfiguration();
+                    config = Xpp3Dom.mergeXpp3Dom(executionConfig, config);
+                    
+                    for (Xpp3Dom elem : config.getChildren()){
+                        if (elem.getName().equals(attribute)) {
+                            return elem.getValue();
+                        }
+                    }
+                    String value = config.getAttribute(attribute);
+                }
+ 
+                break;
+            }
+        }
+        log.debug(attribute + " could not be found in the pom.");
+        return null;
+        
+    }
+
+    private void runTests(String groupId, String artifactId, String phase) throws MojoExecutionException {
+
+        Plugin plugin = project.getPlugin(groupId + ":" + artifactId);
```

Comment:
Can you explain what is happening here before actually running the mojo? Why does the phase need to be specified? Why do you need to delete the previous summary file? What is the purpose of `injectTestId`?

Reply:
> Can you explain what is happening here before actually running the mojo? 

We need to propagate the failsafe/surefire plugin configurations from the existing pom when calling the mojos to run tests.  These configurations can be directly under the `<plugin>`, or it can be under the `<plugin>`'s `<execution>`.  So we need to merge those.

>  Why does the phase need to be specified? 

We call this method using the following, where the third param is the phase:
```
    private void runUnitTests() throws MojoExecutionException {
        runTests("org.apache.maven.plugins", "maven-surefire-plugin", "test");
    }

    private void runIntegrationTests() throws MojoExecutionException {
        runTests("org.apache.maven.plugins", "maven-war-plugin", "war");
        runTests("org.apache.maven.plugins", "maven-failsafe-plugin", "integration-test");
        runTests("org.apache.maven.plugins", "maven-failsafe-plugin", "verify");
    }
```

For `test` phase (unit tests), we inject test ID as described below.
For `integration-test` phase (ITs), we inject test ID as described below, and we also need to cleanup the previous summary file.  Without this, each time we re-run the ITs from the same Maven session, the summary file gets appended instead of replaced, and then the `verify` phase would report results from previous runs.  Therefore we need to clear the summary file before each re-run.

> What is the purpose of injectTestId?

When running tests, the Surefire project's AbstractSurefireMojo calculates a checksum of the plugin parameters to determine whether it should re-run the tests.  If we simply call that mojo again with the exact same parameters in the same Maven session, it will only run tests on the first call and skip every time after that.  We inject an increasing test ID as a dummy property to force AbstractSurefireMojo to calculate a new checksum, allowing the tests to re-run every time.

Reply:
> We call this method using the following, where the third param is the phase:

So you don't actually need the phase to inject the test ID/cleanup the summary file right? You can just rely on the name of the goal. 

It looks like the other thing you are using the phase for is to apply the user's configuration of that phase to the corresponding goal, is that correct? If so, it seems like you expect that these goals to always be mapped to their default phase, which I'm not sure will always be the case. 

If that is the only other reason you need the phase, then I believe you can read the plugin's configuration directly using a method like `MavenProject. getGoalConfiguration()`, such as what is done in `MavenProjectUtil.getPluginConfiguration()`. I am hoping that can eliminate the need for a phase specification.

Reply:
> So you don't actually need the phase to inject the test ID/cleanup the summary file right? You can just rely on the name of the goal.

The test ID is not based on the phase name, but is just a counter.
We need the phase to determine if it's `integration-test` or `verify` phase.  Of course for `verify`, we should not clean up the summary file beforehand.

> It looks like the other thing you are using the phase for is to apply the user's configuration of that phase to the corresponding goal, is that correct? If so, it seems like you expect that these goals to always be mapped to their default phase, which I'm not sure will always be the case.

Correct (although here we are actually using it as the execution id.  I can rename the variable.) It may not always be the case, but this will cover the default cases, and I don't see a way to handle arbitrary user-defined execution ids.  At least when the user runs tests from Maven normally (even without dev mode), they would expect the default execution id to be used.

> If that is the only other reason you need the phase, then I believe you can read the goal's configuration directly using a method like MavenProject.getGoalConfiguration(), such as what is done in MavenProjectUtil.getPluginConfiguration(). I am hoping that can eliminate the need for a phase specification.

I tried that but it needs the specific execution id to get the execution configuration.  Without that or if the execution id doesn't exist in the project, `MavenProject.getGoalConfiguration()` just returns the plugin configuration which isn't useful.  So regardless, we need the logic written here to look for some default execution ids.

Reply:
Refactored it a bit so the code is clearer: https://github.com/WASdev/ci.maven/pull/476/commits/1b470f8d2a47f7e932c988d18d6f6f44ad29d9fc